### ISSUE TITLE
opt: assuming statistical independence due to lack of multi-column stats can cause bad plans

### ISSUE THREAD

**Issue by rytaft** [cockroachdb/cockroach#91597]
**Describe the problem**

When we do not have multi-column stats on a set of columns, we assume that the columns are completely independent. If the columns are actually correlated, this can cause us to choose a bad plan. For example, we have seen that we might choose a zigzag join when it would be much better to perform a filtered scan.

**To Reproduce**

Run the following commands in `./cockroach demo` on master (or 22.2.0):
```
CREATE TABLE tab (a INT, b INT, INDEX (a), INDEX (b));
INSERT INTO tab SELECT 1, 1 FROM generate_series(1, 1000);
INSERT INTO tab SELECT 2, 2 FROM generate_series(1, 1000);
INSERT INTO tab SELECT 3, 3 FROM generate_series(1, 1000);
INSERT INTO tab SELECT 4, 4 FROM generate_series(1, 1000);
ANALYZE tab;
```
We have just created a table in which the two columns, `a` and `b`, are perfectly correlated. However, since there is no index on `(a, b)`, we do not collect multi-column stats on `a` and `b`, so we currently have no way to know they are correlated. Here is what the stats look like:
```
> SHOW STATISTICS FOR TABLE tab;
  statistics_name | column_names |          created           | row_count | distinct_count | null_count | avg_size |    histogram_id
------------------+--------------+----------------------------+-----------+----------------+------------+----------+---------------------
  __auto__        | {a}          | 2022-11-09 14:13:15.857178 |      4000 |              4 |          0 |        2 | 812426185479258113
  __auto__        | {b}          | 2022-11-09 14:13:15.857178 |      4000 |              4 |          0 |        2 | 812426185485320193
  __auto__        | {rowid}      | 2022-11-09 14:13:15.857178 |      4000 |           4000 |          0 |        9 | 812426185472737281
```
Given these stats, the optimizer assumes that `a` and `b` are uncorrelated. Therefore, it estimates the best plan for a query selecting a single value for each column is a zigzag join:
```
> EXPLAIN ANALYZE SELECT * FROM tab WHERE a = 1 AND b = 1;
                          info
--------------------------------------------------------
  planning time: 1ms
  execution time: 192ms
  distribution: full
  vectorized: true
  rows read from KV: 2,999 (190 KiB, 1,002 gRPC calls)
  cumulative time spent in KV: 187ms
  maximum memory usage: 70 KiB
  network usage: 0 B (0 messages)

  • zigzag join
    nodes: n1
    actual row count: 1,000
    KV time: 187ms
    KV contention time: 0µs
    KV rows read: 2,999
    KV bytes read: 190 KiB
    KV gRPC calls: 1,002
    estimated row count: 250
    pred: (a = 1) AND (b = 1)
    left table: tab@tab_a_idx
    left columns: (a, rowid)
    left fixed values: 1 column
    right table: tab@tab_b_idx
    right columns: (b, rowid)
    right fixed values: 1 column
(25 rows)


Time: 195ms total (execution 194ms / network 0ms)
```
Notice that this took `195ms` to run. A much better plan would avoid the zigzag join, which we can see by forcing NO_ZIGZAG_JOIN:
```
> EXPLAIN ANALYZE SELECT * FROM tab@{NO_ZIGZAG_JOIN} WHERE a = 1 AND b = 1;
                                          info
----------------------------------------------------------------------------------------
  planning time: 977µs
  execution time: 6ms
  distribution: full
  vectorized: true
  rows read from KV: 4,000 (164 KiB, 1 gRPC calls)
  cumulative time spent in KV: 5ms
  maximum memory usage: 200 KiB
  network usage: 0 B (0 messages)

  • filter
  │ nodes: n1
  │ actual row count: 1,000
  │ estimated row count: 250
  │ filter: (a = 1) AND (b = 1)
  │
  └── • scan
        nodes: n1
        actual row count: 4,000
        KV time: 5ms
        KV contention time: 0µs
        KV rows read: 4,000
        KV bytes read: 164 KiB
        KV gRPC calls: 1
        estimated max memory allocated: 200 KiB
        estimated row count: 4,000 (100% of the table; stats collected 13 minutes ago)
        table: tab@tab_pkey
        spans: FULL SCAN
(27 rows)


Time: 8ms total (execution 7ms / network 1ms)
```
It only took 8ms to run -- over 10x faster!

**Expected behavior**
We should have some way to represent the uncertainty of our stats when we don't have multi-column stats available for a set of columns. Either we should always assume some level of correlation between columns, or we should represent this possibility with a confidence interval. This will prevent us from choosing such bad plans.

**Environment:**
 - CockroachDB version: 22.2.0
 - Server OS: MacOS
 - Client app: `cockroach demo`

**Additional context**
Related to https://github.com/cockroachlabs/support/issues/1897.

**Comment by msirek** [comment]
It's also a possibility that the gRPC call overhead costs are underestimated, even after being increased by #89261, and the estimated lookup costs of zig-zag join need to be further increased.  The bad plan has:
```
estimated row count: 250
```
... which is 1/4 of the number of gRPC calls in the plan. Even if the actual execution had 250 gRPC calls (and take between 45 and 50 ms), this would still be more expensive than the good plan.  Increasing lookup costs would mean picking zig-zag join in less cases, but maybe this is warranted.  There is also the issue of estimating latency costs.  Right now we don't measure the actual latency between nodes, so can't accurately model the lookup costs (though that might not play a large factor in the test case in this issue).

**Comment by rytaft** [comment]
Good point, @msirek -- the cost of zigzag joins does still seem to be underestimated (although I'm not sure why you think it has to do with gRPC costs? This example was entirely in a single-node `cockroach demo` instance). Please feel free to open up a separate issue if you'd like.

**Comment by msirek** [comment]
> I'm not sure why you think it has to do with gRPC costs?

Well, there is more than just gRPC costs... all of the costs associated with starting a new scan.  Just observing that the number of gRPC may be roughly equal or proportional to the number of these new scan starts, and doing this roughly once for every qualified row is expensive. Costs which only occur once for the entire operation in the single scan, single gRPC competing plan.

I see in lookup join costs, extra costs added when the lookup columns don't form a key:
https://github.com/cockroachdb/cockroach/blob/3ab4b23e6a397b2237047ad05a8aea47844e39e7/pkg/sql/opt/xform/coster.go#L1005-L1013
The comment mentions these costs no longer apply with streaming lookup join, but I don't think zig zag join uses the streamer, so we may want to add similar costs to zig zag join.
